### PR TITLE
Add `kManagerDriven` and `kRequired` access modes

### LINKED ISSUE #1209: Add `kManagerDriven` and `kRequired` to `PolicyAccess`
(Capturing some nascent rumination on issues that came up working on #31) 

## What

- Add `kManagerDriven` and `kRequired` to `PolicyAccess` for when a Host wishes to determine which traits a Manager requires for publishing or can provide for a future entity.
- Remove `ResolvesFutureEntitiesTrait` from MediaCreation

Such that, from the Hosts perspective, `PolicyAccess` cases are defined as follows for _entity_ trait sets:

- `kRead`: What is your policy for existing entities with the supplied trait set, and which traits are you capable of resolving data for.
- `kWrite`: What is your policy when I wish to publish entities with the supplied trait set, that I already hold the data for, and which traits are you capable of storing data for?
- `kManagerDriven`: What is your policy when I wish to publish entities with the supplied trait set, and which of the traits are you capable of providing data for (via resolve) for the yet-to-be-published entity. 
- `kRequired`: What is your policy when I wish to publish entities with the supplied trait set, and which of the traits are required.
- `kCreateRelated`: As per `kWrite` (included for parity with `PublishAccess`.

## Why

Managers may required the data for certain properties to be present in order to publish an entity.  The core API also defines the `prefilght` -> `resolve` -> `register` workflow for when a Host wishes or requires to derive some of an entities properties from the Manager during publish. For example - the location to store a render, or what frame range to export.

There is however, no way for a Host to query ahead of time which traits the Manager requires or may be able to resolve for future entities. This means that the Host has no way to present a coherent user experience without "trying first". The only indication that a Manager is even capable of resolving future data is via the tenuous [`ResolvesFutureEntities`](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/blob/4f73fb6aa062d5807c4eb8010ff792da391b7d4b/traits.yml#L221C7-L221C29) trait from the optional and distinct MediaCreation library, which provides no information about which specific traits are derivable.

We need a solution within the core API, given these are are core workflows.

Using `managementPolicy` to provide this data has the benefit that as the method does not take an Entity Reference, it is a broad application-level query which allows Hosts to preemptively manage user experience if they are not able to provide data a manager needs or the manager isn't able to provide data the host needs. The resulting low call count hopefully offsets that several calls may be needed to define final behaviour.

It also constrains this business logic to specific entity trait sets, which simplifies implementation as it removes the need (or possibility) for this to vary between entities of the same trait set, which feels beneficial in what is an already complex working environment.

It would be left to Trait definitions to define which particular properties are required or optional. A trait being indicated as supported/required/derivable by `mangementPolicy` would, by contract, required that Hosts provide those properties, and/or Managers would populate them during `resolve`.

## Notes

There are some other potential ideas to help with other areas:

### `ResolveAccess`

`ResolveAccess::kWrite` could be changed to `ResolveAccess::kManagerDriven`, as resolving for write is always the `kManagerDriven` policy case (as if the Host already has the data it wouldn't resolve it). This would be a breaking change though.

### Preflight help

An extension of this could be to add `kWrite_Collaborative` and `kCreateRelated_Collaborative` to `PublishingAccess` for when a Host will attempt to resolve derivable traits from the `preflight` reference. This would help inform the business logic in `preflight` to validate the input. It doesn't completely solve the problem though as it doesn't _guarantee_ that the Host would try to resolve _all_ the missing required properties. So it might not be worth complicating the `PublishingAcces` enum given its not a robust solution. 

An alternative could be to add an optional trait set to `preflight` that indicates which traits will be resolved during publishing. However this list only improves validation, as the host still woudln't know which traits a manager might be able to resolve without the above proposed mechanism, and it woudln't provide means to refine that list beforehand. It also has complexities when publishing multiple entity trait sets in one batch.

It would perhaps be more robust (and less intrusive) to add method(s) to pre-validate the TraitsData (that would be) passed to `preflight` instead, that return programmatically inspectable failure descriptions (this allows value ranges to be checked as well as just missing data) this, when also supplied with a list of "traits Im going to resolve during publish" it would perhaps be the most effective way to pro-actively defend against "try it an see if it explodes" workflows in relation to specific entity references and planed publishing approaches. Being entity-specific though, so is not suitable for high-level application behaviour switching so is probably not an alternative to the above proposed `managementPolicy` approach. This would also need consideration in relations to the "permissions" mechanism.

### Alternatives

 - Splitting `managementPolicy` into different methods eg: `managementPolicy` and `publishingPolicy` that simplifies the Access enums and makes it more obvious.  A much bigger scope of change though.

## ACs
```[tasklist]
- [ ] Add `kManagerProvided` and `kRequired` to `PolicyAccess`
- [x] Remove `ResolvesFutureEntitiesTrait` from MediaCreation (might already be done)
- [ ] https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/issues/98
- [ ] https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/issues/75
- [ ] Scan docs and update any publishing examples
- [ ] Update existing publishing examples to account for kRequired/kManagerProvided
```

### PR DESCRIPTION
**Pull request by feltech**
## Description

Closes #1209

Add the `kRequired` and `kManagerDriven` access modes for use with `managementPolicy`; and rename `ResolveAccess.kWrite` to `kManagerDriven`, for consistency.

- [x] I have updated the release notes.
- [x] I have updated all relevant user documentation.

## Test Instructions

See
- https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/76
- https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/99

### CODE REVIEW

**Review by elliotcmorris — CHANGES_REQUESTED** [review]
Looks grand. Only minor doc comments.

File: src/openassetio-core/include/openassetio/access.hpp [hunk 1]
```diff
@@ -49,6 +50,22 @@ enum class PolicyAccess : std::underlying_type_t<internal::access::Access> {
    * @ref RelationsAccess.kCreateRelated
    */
   kCreateRelated = internal::access::Access::kCreateRelated,
+  /**
+   * Host wishes to know which subset of traits must have their required
+   * properties filled for successful publishing of an entity.
+   *
+   * @see @ref PolicyAccess.kRequired
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/access.hpp:57]: Whoops, good spot

File: src/openassetio-core/include/openassetio/access.hpp [hunk 2]
```diff
@@ -49,6 +50,22 @@ enum class PolicyAccess : std::underlying_type_t<internal::access::Access> {
    * @ref RelationsAccess.kCreateRelated
    */
   kCreateRelated = internal::access::Access::kCreateRelated,
+  /**
+   * Host wishes to know which subset of traits must have their required
+   * properties filled for successful publishing of an entity.
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/access.hpp:55]: I don't think we have any better place to elaborate on this, so I can just expand the docs here.

File: src/openassetio-core/include/openassetio/hostApi/Manager.hpp [hunk 3]
```diff
@@ -373,9 +373,8 @@ class OPENASSETIO_CORE_EXPORT Manager final {
    *  - For @fqref{access.PolicyAccess.kRead} "read" usage, any
    *    additional traits with properties that you wish to resolve for
    *    that type of entity.
-   *  - For @fqref{access.PolicyAccess.kWrite} "write" usage, any
-   *    additional traits with properties that you wish to publish for
-   *    that type of entity.
+   *  - For publishing usage, any additional traits with properties that
+   *    you wish to publish for that type of entity.
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/hostApi/Manager.hpp:377]: I don't think that's right. This is the doc for the `managementPolicy`, so providing the  `kRequired` traits requires querying the policy in the first place, only to query the policy again with those traits?

Regardless, the `managementPolicy` docs could do with expanding to make more sense for the new access modes.

File: RELEASE_NOTES.md [hunk 4]
```diff
@@ -12,6 +12,11 @@ v1.0.0-beta.x.x
   must have, for a given entity reference.
   [#31](https://github.com/OpenAssetIO/OpenAssetIO/issues/31)
 
+- Access policy for resolving future entities, `ResolveAccess.kWrite`
+  has been renamed to `ResolveAccess.kManagerDriven`, for consistency
+  with the `managementPolicy` access mode of the same name.
+  [#1209](https://github.com/OpenAssetIO/OpenAssetIO/issues/1209)
```
[the remark this answers is not in the document]
Comment (feltech) [reply RELEASE_NOTES.md]: Reworded slightly in b88faff1

File: doc/doxygen/src/Examples.dox [hunk 5]
```diff
@@ -380,25 +384,21 @@
...
- *     if save_url := working_spec.locatableContentTrait().getLocation():
- *       save_path = pathFromURL(save_url)  # URL decode etc
- *     if custom_encoding := working_spec.textEncodingTrait().getEncoding():
- *       encoding = custom_encoding
- *   except BatchElementException as exc:
- *     # Only ignore the case the manager can't provide data
- *     if exc.error.code != exc.error.ErrorCode.kEntityAccessError
- *        raise exc
+ * # If the manager wants to drive at least one of the traits we're
+ * # publishing, we can `resolve` in `kManagerDriven` mode to
```
[the remark this answers is not in the document]
Comment (feltech) [reply doc/doxygen/src/Examples.dox]: Tweaked as suggested in 9bcc8550

File: src/openassetio-core/include/openassetio/access.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/access.hpp:55]: Elaborated in 060c8720. Also see updates in response to https://github.com/OpenAssetIO/OpenAssetIO/pull/1255#discussion_r1475993032

File: src/openassetio-core/include/openassetio/access.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/access.hpp]: Removed in 060c8720

File: src/openassetio-core/include/openassetio/hostApi/Manager.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply src/openassetio-core/include/openassetio/hostApi/Manager.hpp:386]: :facepalm: Oh yeah you're right. I got it in my head this was talking about a publishing action post managementPolicy.

**Review by elliotcmorris — APPROVED** [review]
LGTMSAMWR

**Pull request merged**